zapgrpc/zapgrpc.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package zapgrpc provides a logger that is compatible with grpclog.
+package zapgrpc // import "go.uber.org/zap/zapgrpc"
+
+import "go.uber.org/zap"
+
+// LoggerOption is an option for a new Logger.
+type LoggerOption func(*loggerOptions)
+
+// WithDebug says to print the debug level instead of the default info level.
+func WithDebug() LoggerOption {
+	return func(loggerOptions *loggerOptions) {
+		loggerOptions.printFunc = (*zap.SugaredLogger).Debug
+		loggerOptions.printfFunc = (*zap.SugaredLogger).Debugf
+	}
+}
+
+// NewLogger returns a new Logger.
+func NewLogger(l *zap.Logger, options ...LoggerOption) *Logger {
+	loggerOptions := newLoggerOptions(options...)
+	return &Logger{
+		l.Sugar(),
+		(*zap.SugaredLogger).Fatal,
+		(*zap.SugaredLogger).Fatalf,
+		loggerOptions.printFunc,
+		loggerOptions.printfFunc,
+	}
+}
+
+// Logger is a logger that is compatible with the grpclog Logger interface.
+type Logger struct {
+	sugaredLogger *zap.SugaredLogger
+	fatalFunc     func(*zap.SugaredLogger, ...interface{})
+	fatalfFunc    func(*zap.SugaredLogger, string, ...interface{})
+	printFunc     func(*zap.SugaredLogger, ...interface{})
+	printfFunc    func(*zap.SugaredLogger, string, ...interface{})
+}
```

Comment:
What's the benefit of keeping these four function pointers instead of

```
type Logger struct {
  sugar *zap.SugaredLogger
  development bool
}
```

and checking `development` in `Print` and friends? I don't think that the performance impact of a single boolean check will even be measurable relative to the cost of string formatting.

This also lets us get rid of the the `loggerOptions` type.